### DISCUSSION TITLE
DS3231 - get_ambient_temperature()

### DISCUSSION THREAD

**Discussion by PaulBags** [astlouys/Pico-Green-Clock#25] [Q&A]
I've been digging through code and the [data sheet](https://www.analog.com/en/products/ds3231.html#product-evaluationkit) for the DS3231 chip, and I have some concerns about the get_ambient_temperature() function - and some confusion.

Firstly my concern, by writing {0x0E, 0x20} it looks like we're writing the entire control byte - assuming the data sheet is correct this should have several side effects:
- Setting bit 7 to 0 (default) enables the oscillator while on battery.
- Setting bit 6 to 0 (default) as well as bit 2 to 0 (not default) disables the square wave while voltage is too low (might not be on battery).
- Setting bits 4,3 to 0,0 (not default) sets square wave frequency to 1Hz.
- Setting bit 2 to 0 (not default) enables square wave output.
- Setting bits 1,0 to 0,0 (default) disables alarm 2 and 1 interrupts.
Setting bit 5 to 1 is.. interesting. While it does flush temperature to data, it _also_ forces TCXO execution (data sheet, pg. 13) - this doesn't appear to be an instantaneous process, nor is it supposed to be called when the chips internal TCXO execution cycle is running.

While I'm not sure that any of these are bugs per se since pico green clock does what it intends to, anyone who wants to modify code to work more closely with the DS3231 chip might run into problems - because get_ambient_temperature() is doing a lot more than just getting the ambient temperature.

Now to my confusion. I'm still trying to wrap my head around C, but, writing &GetAddHigh and &GetAddLow (GetAddHigh and GetAddLow are UCHAR, not pointers) appears to be using the local memory addresses as registers to access rather than 0x11 and 0x12 (the upper and lower temp registers on the DS3231). This can't be right? And yet the temperature read always appears to be sane, if not particularly accurate.

**Comment by astlouys** [comment]
Hello Paul,

This is the first time I receive comments about the DS3231... That's interesting.

I won't have time to take a look at that in the very next days, but I'm curious to check your observations with more details. While working on the code, I remember having analyzed alarms with some care, but nothing related to what you mention...

to be continued...
